### PR TITLE
Add support for using a HTTP proxy for outbound connections (just for app install for now)

### LINKED ISSUE #693: HTTP Proxy for outbound requests
Further to the email discussion with @kentonv:

> Currently there are several places in Sandstorm's code where we use Node's "http" and "https" modules to make outgoing HTTP requests.
> 
> It looks like the following files use these modules:
>
> shell/server/sandcats.js: To talk to the Sandcats server for dynamic DNS updates, if you use sandcats.
> shell/server/installer.js: To download packages.
> shell/server/hack-session.js: To implement the httpGet() API method, which e.g. TinyTinyRSS uses to fetch RSS feeds.
> shell/server/drivers/external-ui-view.js: To implement a more general mechanism for apps to make outgoing HTTP requests (not yet used by any apps).
>
> And this file uses Meteor's `HTTP` API:
>
> shell/packages/sandstorm-db/user.js: To fetch profile pictures from Google and Github.
>
> Meteor's HTTP is a wrapper around the `request` npm package, which in turn wrap's node's HTTP client. It looks like `request` explicitly supports proxies (and Meteor's HTTP supports passing arbitrary options to `request`):
>
> https://www.npmjs.com/package/request#proxies
>
> It's also possible to get Node's raw HTTP client to use proxies per this StackOverflow question:
>
> http://stackoverflow.com/questions/3862813/how-can-i-use-an-http-proxy-with-node-js-http-client
>
> So we need to make sure that all the places where we make HTTP requests use the proxy.
>
> I think the right way to do this would be to introduce a new Sandstorm-internal HTTP API that either wrap's Meteor's `HTTP` or the npm `request` package (probably the latter) and allows us to force all requests globally through a porxy, configurable through the admin interface. Then we should update all the above code to use it.
>
> Additionally, run-bundle.c++ shells out to `curl` to fetch Sandstorm updates. I believe you can make `curl` use a proxy by setting an environment variable, so if you set that variable in the initscript that calls `sandstorm start` it should apply. (Maybe it makes sense for the Sandstorm shell to pick up the http_proxy environment variable automatically as well, rather than require setting it in the admin settings?)

### PR DESCRIPTION
**Pull request by paulproteus**
This adjusts `installer.js` to use the `request` API, rather than the node
HTTP/HTTPS API. It also adjusts `run-bundle.c++` to pass this variable through,
and adjusts the docs to explain how to use this feature.

Refs #693

Does not fully close #693 yet because a few other places need a similar update.

### CODE REVIEW

File: shell/server/installer.js [hunk 1]
```diff
@@ -310,92 +310,74 @@ AppInstaller = class AppInstaller {
     };
 
     let protocol;
-    if (url.protocol === 'http:') {
-      protocol = Http;
-    } else if (url.protocol === 'https:') {
+    if (url.protocol === 'http:' || url.protocol === 'https:') {
+      protocol = Request.defaults({
+        maxRedirects: 1,
```
Comment (kentonv) [inline shell/server/installer.js]: Why only one? Note that we have historically had app URLs that were redirects, and double-redirects are not that uncommon.

File: shell/server/installer.js [hunk 2]
```diff
@@ -310,92 +310,74 @@ AppInstaller = class AppInstaller {
...
       // CAs available to Sandstorm.
-      options.rejectUnauthorized = false;
-      protocol = Https;
+        strictSSL: false
+      });
     } else {
       throw new Error('Protocol not supported: ' + url.protocol);
     }
 
     // TODO(security):  It could arguably be a security problem that it's possible to probe the
```
Comment (kentonv) [inline shell/server/installer.js]: Hey I think you're fixing this TODO! (Networks with significant private internals should always specify a proxy.)

File: src/sandstorm/run-bundle.c++ [hunk 3]
```diff
@@ -1162,12 +1162,23 @@ private:
 
     // The environment inherited from the host is probably no good for us. E.g. an oddball
     // locale setting can crash Mongo because we don't have the appropriate locale files available.
+    //
+    // That said, there are a few environment variables that we do re-export.
+    char * http_proxy, * https_proxy;
```
Comment (kentonv) [inline src/sandstorm/run-bundle.c++]: Generally in C++ it's recommended to associate the `*` with the type, i.e. `char* x` not `char *x`. This of course leads to confusion when declaring multiple variables at once as you are doing here, and generally the answer to that is: don't do that. :) That is, declare them on separate lines.

File: src/sandstorm/run-bundle.c++ [hunk 4]
```diff
@@ -1162,12 +1162,23 @@ private:
...
+    KJ_SYSCALL(http_proxy = getenv("https_proxy"));
```
Comment (kentonv) [inline src/sandstorm/run-bundle.c++]: `s/https/http/`

Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: +1

File: src/sandstorm/run-bundle.c++ (hunk 3 above)
Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: +1

File: shell/server/installer.js (hunk 1 above)
Comment (paulproteus) [inline shell/server/installer.js]: How about 5? : D

I just want to pick a constant, any constant.

File: src/sandstorm/run-bundle.c++ (hunk 3 above)
Comment (kentonv) [inline src/sandstorm/run-bundle.c++]: Also please use camelCase, not under_scores.

File: shell/server/installer.js (hunk 2 above)
Comment (paulproteus) [inline shell/server/installer.js]: Clearly you've never worked at Eventbrite!

Which is to say - I think there should be some other mechanism for avoiding "server side request forgery". And/or we should document the heck out of this.

File: src/sandstorm/run-bundle.c++ (hunk 4 above)
Comment (kentonv) [inline src/sandstorm/run-bundle.c++]: `KJ_SYSCALL` only works for functions that return `int`. Also it's only meant for calls that can fail, but getenv() cannot fail (it returns null when there's no match, but that's not a failure).

File: src/sandstorm/run-bundle.c++ (hunk 3 above)
Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: Oh, I was mirroring the environment variable name. +1, I can use camelCase for this.

File: src/sandstorm/run-bundle.c++ [hunk 5]
```diff
@@ -1162,12 +1162,23 @@ private:
...
+    KJ_SYSCALL(https_proxy = getenv("https_proxy"));
     KJ_SYSCALL(clearenv());
```
Comment (kentonv) [inline src/sandstorm/run-bundle.c++]: I believe that this `clearenv()` call frees the memory that your `http_proxy` and `https_proxy` pointers are pointing at. You will need to copy the strings, preferably using `kj::str()` -- but of course only if they aren't null.

Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: "it seems to work fine" (but I share your assessment from the man page)

+1 to `kj::str()` to copy

File: shell/server/installer.js (hunk 1 above)
Comment (kentonv) [inline shell/server/installer.js]: Apparently Chrome and Firefox use a limit of 20 by default.

File: src/sandstorm/run-bundle.c++ (hunk 4 above)
Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: Cool, I will try without `KJ_SYSCALL`.

File: shell/server/installer.js (hunk 1 above)
Comment (paulproteus) [inline shell/server/installer.js]: Fixed in latest push

**Comment by paulproteus** [comment]
@kentonv , how do you suggest I express the following?

```
    const char * keepTheseVars[] = {"http_proxy", "https_proxy"};
    auto numberOfVarsToKeep = 2; // FIXME: How am I supposed to express this & the previous line?
    for (int i = 0; i < numberOfVarsToKeep; i++) {
```

Rationale for doing it this way: I had typos in `https_proxy` vs. `http_proxy` earlier, and so I want to do this in a less repeating-myself fashion to avoid future typos and increase clarity. But I don't know how to make a literal string array in KJ/C++.

Other than the above style question, I believe this addresses all review feedback.

**Comment by kentonv** [comment]
`kj::size(keepTheseVars)` will evaluate to `2`.

Also, in C++11 you can write:

    for (char* var: keepTheseVars) {

**Comment by paulproteus** [comment]
Thanks @kentonv 

I've pushed a style fix for that, and additionally I've manually re-tested that app installs work, for both http and https app URLs. Ready for your re-review/merge.

File: src/sandstorm/run-bundle.c++ (hunk 3 above)
Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: Fixed by avoiding having so many temp variables in the first place.

File: src/sandstorm/run-bundle.c++ (hunk 4 above)
Comment (paulproteus) [inline src/sandstorm/run-bundle.c++]: Fixed by not using `KJ_SYSCALL` on getenv, merely on setenv, which can error.

**Comment by paulproteus** [comment]
Pushed a fix to my "style fix" which makes it actually compile. I will let Jenkins test this; anyway, I did also re-manually-test this locally.

**Comment by kentonv** [comment]
I added some OCD style tweaks directly in github. I'll wait for Jenkins to verify I didn't typo anything then merge.

**Pull request merged**